keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+these fields, rather than needing to do a kind-resource lookup for every user's
+interaction with the API.
+
+Secondly, as in the Ingress API, the target types must be clearly defined by
+each implementation, so there should be no opportunity for ambiguity.
+
+Lastly, this is only half of a handshake, so the From referent must also both
+support referencing the To referent, and must reference the one in the To field.
+
+With all of that in mind, we felt that the benefits in terms of specificity
+outweighed the costs in terms of user pain to use the more correct `Resource`.
+
+### API Spec
+
+```golang
+// ReferenceGrant identifies kinds of resources in other namespaces that are
+// trusted to reference the specified kinds of resources in the same namespace
+// as the policy.
+//
+// Each ReferenceGrant can be used to represent a unique trust relationship.
+// Additional ReferenceGrants can be used to add to the set of trusted
+// sources of inbound references for the namespace they are defined within.
+//
+// ReferenceGrant is a form of runtime verification allowing users to assert
+// which cross-namespace object references are permitted. Implementations that
+// support ReferenceGrant MUST NOT permit cross-namespace references which have
+// no grant, and MUST respond to the removal of a grant by revoking the access
+// that the grant allowed.
+//
+// Support: Core
+type ReferenceGrant struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ReferenceGrant.
+	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+
+	// Note that `Status` sub-resource has been excluded at the
+	// moment as it was difficult to work out the design.
+	// A `Status` sub-resource may be added in the future.
+}
+
+// +kubebuilder:object:root=true
+// ReferenceGrantList contains a list of ReferenceGrant.
+type ReferenceGrantList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ReferenceGrant `json:"items"`
+}
+
+// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
+// for Gateway API.
+type ReferenceGrantSpec struct {
+	// From describes the trusted namespaces and kinds that can reference the
+	// resources described in "To". Each entry in this list MUST be considered
+	// to be an additional place that references can be valid from, or to put
+	// this another way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	From []ReferenceGrantFrom `json:"from"`
+
+	// To describes the resources that may be referenced by the resources
+	// described in "From". Each entry in this list MUST be considered to be an
+	// additional place that references can be valid to, or to put this another
+	// way, entries MUST be combined using OR.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
```

Comment:
In general, we've tried to err on the side of having _some_ limit for all lists in Gateway API objects, to ensure that no-one runs up against etcd object size limits. For the From and To fields here, we're trying to encourage people to have more, smaller ReferenceGrant objects, so that we don't end up with single objects granting access to many things.

tbh I don't remember why we picked 16 over some other number though.

Reply:
you;ve got to start somewhere :)

Reply:
Every list field should have a defined limit.

Every string field should have a max size.